common/txmgr/tracker.go

```diff
@@ -0,0 +1,295 @@
+		return fmt.Errorf("failed to get non fatal txes from txStore: %w", err)
+	}
+
+	// insert abandoned txes
+	for _, tx := range nonFatalTxes {
+		if !tr.enabledAddrs[tx.FromAddress] {
+			tr.insertTx(tx)
+		}
+	}
+
+	if err := tr.handleTxesByState(ctx, 0); err != nil {
+		return fmt.Errorf("failed to handle txes by state: %w", err)
+	}
+
+	return nil
+}
+
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) HandleTxesByState(ctx context.Context, blockHeight int64) error {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	return tr.handleTxesByState(ctx, blockHeight)
+}
+
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleTxesByState(ctx context.Context, blockHeight int64) error {
+	for id, atx := range tr.txCache {
+		tx, err := tr.txStore.GetTxByID(ctx, atx.id)
+		if err != nil {
+			return fmt.Errorf("failed to get tx by ID: %w", err)
+		}
+
+		switch tx.State {
+		case TxConfirmed:
+			if err := tr.handleConfirmedTx(tx, blockHeight); err != nil {
+				return fmt.Errorf("failed to handle confirmed txes: %w", err)
+			}
+		case TxConfirmedMissingReceipt, TxUnconfirmed:
+			// Keep tracking tx
+		case TxInProgress, TxUnstarted:
+			// Tx could never be sent on chain even once. That means that we need to sign
+			// an attempt to even broadcast this Tx to the chain. Since the fromAddress
+			// is deleted, we can't sign it.
+			errMsg := "abandoned transaction could not be sent on chain"
+			if err := tr.markTxFatal(ctx, tx, errMsg); err != nil {
+				return fmt.Errorf("failed to mark tx as fatal: %w", err)
+			}
+			delete(tr.txCache, id)
+		case TxFatalError:
+			delete(tr.txCache, id)
+		default:
+			tr.lggr.Errorw(fmt.Sprintf("unhandled transaction state: %v", tx.State))
+		}
+	}
+
+	return nil
+}
+
+// handleConfirmedTx removes a transaction from the tracker if it's been finalized on chain
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleConfirmedTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	blockHeight int64,
+) error {
+	finalized, err := tr.txStore.IsTxFinalized(context.Background(), blockHeight, tx.ID)
+	if err != nil {
+		return fmt.Errorf("failed to check if tx is finalized: %w", err)
+	}
+
+	if finalized {
+		delete(tr.txCache, tx.ID)
+	}
+
+	return nil
+}
+
+// insertTx inserts a transaction into the tracker as an AbandonedTx
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) insertTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	if _, contains := tr.txCache[tx.ID]; contains {
+		return
+	}
+
+	tr.txCache[tx.ID] = AbandonedTx[ADDR]{
+		id:          tx.ID,
+		fromAddress: tx.FromAddress,
+	}
+	tr.lggr.Debugw(fmt.Sprintf("inserted tx %v", tx.ID))
+}
+
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) markTxFatal(ctx context.Context,
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	errMsg string) error {
+	tx.Error.SetValid(errMsg)
+
+	tx.State = TxInProgress
```

Comment:
Noticed another comment on the tx state to use so this is only relevant if we decide to keep `fatal_error`. Could you add a comment here to explain why you have to set it to `TxInProgress`? Reading the logic for the txstore method would explain it but think it'd be convenient to have a one liner here.